rfcs/0012-declarative-virtual-machines.md

```diff
@@ -0,0 +1,249 @@
+This RFC describes a way to declaratively add virtual machines to NixOS
+machines, in a way similar to the current `containers` module.
+
+# Motivation
+[motivation]: #motivation
+
+## Issues with containers
+[issues-with-containers]: #issues-with-containers
+
+The `containers` module is useful, but is only namespace-level virtualisation.
+As a natural consequence, it blocks kernel-level virtualisation, thus limiting
+the security benefits.
+
+Moreover, the nix store is shared with the host, which means secrets potentially
+put there by the host (and with [issue 8](https://github.com/NixOS/nix/issues/8)
+these can easily come unannounced) are readable from the guest.
+
+Worse, even assuming [issue 8](https://github.com/NixOS/nix/issues/8) is solved,
+the guest is still able to get the host's configuration by reading it from the
+store. This information leak is precious to an attacker trying to attack the
+host system.
+
+In addition, containers have seen many high-importance issues and there are ways
+to escape them (see CVE-2016-5195 or CVE-2017-6074 for recent ones in the linux
+kernel, unfortunately systemd seems not to keep a record of what CVEs were fixed
+in their NEWS file). This is normal, as the attack surface of a container is
+much larger than the one of an emulator, even when considering virtio drivers.
+
+## Use case
+
+The use case this RFC puts forward is the the same as the one of the current
+`containers.` module, but adapted for the needs of someone for whom security is
+more important than speed (pushing for VMs instead of containerization), yet who
+want the same ease of use as with containers.
+
+## Expected outcome
+
+This RFC describes a prospective NixOS module for running VMs. Thus, the main
+design point is that a `vms.` module should be made available to NixOS users.
+It should allow to declaratively define virtual machines, including their
+configuration.
+
+# Detailed design
+[design]: #detailed-design
+
+The VM module is architectured in the following way:
+ * The root filesystem is written on a qcow2 image
+ * The store is stored on the host, for easy upgrade handling (see [Disk
+   management](#disk-management))
+ * Each VM is run as a (set of) systemd service, and can be rebooted using
+   `systemctl restart vm-${name}.service`
+ * `qemu` fetches the kernel and initrd directly from the guests' store
+ * Networking is done with all the VMs in a bridge internal to the server (see
+   [Networking](#networking))
+
+Details follow.
+
+## Basic configuration
+
+The user should be able to configure various aspects of the VMs running,
+especially disk, memory and vcpu usages.
+
+### Proposed option set
+
+The following options are proposed:
+
+```nix
+{
+  vms = {
+    path = "/path/to/dir"; # Path into which to store persistent data (disk
+                           # images and per-vm store)
+    rpath = "/runtime/dir"; # Path for temporary non-user-facing low-size data,
+                            # like IPC sockets
```

Comment:
Renamed `socketDir` as in the current implementation it contains only sockets :)